### ISSUE TITLE
Model training

### ISSUE THREAD

**Issue by theElandor** [nnistelrooij/3dteethland#4]
Dear Niels,
First of all, congratulations on the excellent work, your pipeline is really well designed.
We’re interested in predicting a few specific dental landmarks that weren’t included in your original work, so I’m planning to fine-tune the landmark model to adapt it to this new task.
Since there are no training instructions available, I was wondering if you could please share some guidance on the following:
  1) How should the dataset folder be structured?
  2) How would the configuration file look like for this setup?
  3) Do you think it's a good idea to finetune instead of training from scratch?
Thanks in advance, keep up the good work!
Matteo Lugli

**Comment by nnistelrooij** [comment]
Dear Matteo,

Thank you for your enthusiasm!

1. The dataset follows the structure of the datasets for the [3DTeethSeg](https://osf.io/5cmg3/overview) and [3DTeethLand](https://osf.io/um96h/overview) datasets. You can put all your cases with tooth segmentations in one root folder. This root folder has one folder for each case and each of these case folders should have the upper and lower meshes with `_lower` and `_upper` in the filename and corresponding tooth segmentations as JSON files with the same file name (except the suffix).
For the landmark annotations, you should have a separate landmarks root folder that has landmarks annotations again with corresponding file names and `_lower__kpt.json` or `_upper__kpt.json` as suffix.
These can be specified in the configuration using the `root` and `landmarks_root` variables.
2. If you use the same number and names for the landmarks, the only things you should have to change are the `root`, `landmarks_root`, and `work_dir` parameters.
3. It is always good to at least try! I implemented the landmark detection as a regression problem (offset prediction), so the general knowledge for this task can be transferred when fine-tuning the model. As such, the training should at least be more stable during the first few epochs.

Once you have prepared the data, and still experience problems, please let me know!

Niels

**Comment by theElandor** [comment]
Hello Niels,
I'll try again as soon as possible.
I am working on the prediction of a new dental landmark, so I guess I'll need to do minimal changes to your pipeline.
Thanks for the clarification!
Have a good day,
ML

**Comment by DoubleXIV** [comment]
Hello, I am using your checkpoint model for inference, and some small pieces of gum connected to the crown are being recognized as part of the crown. Is this an issue with the instseg model or the landmarks model? If I want to improve this problem, which model do I need to retrain?

**Comment by theElandor** [comment]
This appears to be an issue with the segmentation module. You may want to try cleaning the scan or rotating it so that it more closely matches the training data distribution.

I’ve used his model for inference on many scans, and it generally performs very well.

If you’re interested, I also have a production-ready intraoral scan segmenter that I trained using Niel’s segmentations as soft labels. It’s built on the Pointcept framework and is extremely fast and reliable for inference. Let me know if that could be helpful.

Have a nice day 🙂

**Comment by czc98** [comment]
@theElandor Hello, I tried to run the inference steps mentioned in the readme, but it failed due to some missing model checkpoints (.ckpt files). If I want to successfully run infer.py, or run the segmentation and landmark detection separately, what should I do? Could you please give me some advice? Thank you.

**Comment by DoubleXIV** [comment]
Hello, where should I go to obtain this production-ready intraoral scan segmenter? I would be very grateful if you could provide it to me.

**Comment by theElandor** [comment]
I am working in these days to clean up the code and publish the repo. It's based on the [pointcept](https://github.com/Pointcept/Pointcept) framework, so it should be relatively easy to run.

**Comment by theElandor** [comment]
> [@theElandor](https://github.com/theElandor) Hello, I tried to run the inference steps mentioned in the readme, but it failed due to some missing model checkpoints (.ckpt files). If I want to successfully run infer.py, or run the segmentation and landmark detection separately, what should I do? Could you please give me some advice? Thank you.

This is the config file we used. Sorry for the absolute paths, but the checkpoint names should be the same.
```yaml
work_dir: '/work/grana_maxillo/Mlugli/'
version: mixedseg_validate
seed: 0

model:
  align:
    channels_list: [48, 96, 192, 256]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.4, 0.8, 1.6]
    point_embedding:
      use: True
      kpconv_point_influence: 0.04
      kpconv_ball_radius: 0.1
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    checkpoint_path: '/home/mkaailab/Documents/IOS/partials/checkpoints/china_align_all_0.ckpt'
    checkpoint_path: 'checkpoints/3dteethseg_align_all.ckpt'
    checkpoint_path: '/mnt/diag/IOS/partials/checkpoints/3dteethseg_align_all.ckpt'
    lr: 0.001
    weight_decay: 0.0001
    epochs: 1000
    warmup_epochs: 5
    do_seg: True

  instseg:
    channels_list: [48, 96, 192, 256]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.4, 0.8, 1.6]
    point_embedding:
      use: True
      kpconv_point_influence: 0.04
      kpconv_ball_radius: 0.1
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    checkpoint_path: '/work/grana_maxillo/BiteClassify/instseg_full.ckpt'
    # checkpoint_path: '/home/mkaailab/Documents/logs/tooth_segmentation/mixedseg_checked/checkpoints/weights-epoch=132-v1.ckpt'
    lr: 0.0006
    weight_decay: 0.0001
    epochs: 200  # overfitting after 600
    warmup_epochs: 5

  mixedseg:
    channels_list: [48, 96, 192, 256]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.4, 0.8, 1.6]
    point_embedding:
      use: True
      kpconv_point_influence: 0.04
      kpconv_ball_radius: 0.1
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    checkpoint_path: '/work/grana_maxillo/BiteClassify/landmarks_full.ckpt'
    lr: 0.0006
    weight_decay: 0.0001
    epochs: 200
    warmup_epochs: 0

  landmarks:
    channels_list: [48, 96, 192, 256]
    out_channels: [1, 4, 4, 4, 4, 4]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.1, 0.2, 0.4]
    point_embedding:
      use: True
      kpconv_point_influence: 0.02
      kpconv_ball_radius: 0.05
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    # checkpoint_path: 'checkpoints/landmarks.ckpt'
    checkpoint_path: '/work/grana_maxillo/BiteClassify/landmarks_full.ckpt'
    lr: 0.0006
    weight_decay: 0.0001
    epochs: 500
    warmup_epochs: 5
  
  binseg:
    channels_list: [48, 96, 192, 256]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.1, 0.2, 0.4]
    point_embedding:
      use: True
      kpconv_point_influence: 0.015
      kpconv_ball_radius: 0.03
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    checkpoint_path: '/home/mkaailab/Documents/IOS/partials/checkpoints/china_binseg_partialnoalign_4.ckpt'
    checkpoint_path: '/mnt/diag/IOS/Maud Wijbrandts/checkpoints/maud_lu_c_fractures_0/checkpoints/weights-epoch=79.ckpt'
    checkpoint_path: '/mnt/diag/IOS/partials/checkpoints/3dteethseg_binseg_all.ckpt'
    checkpoint_path: '/home/mkaailab/Documents/logs/tooth_segmentation/zainab2_foldall_dicefocal_pretrain/checkpoints/weights-epoch=199.ckpt'
    lr: 0.0006
    weight_decay: 0.0001
    epochs: 200
    warmup_epochs: 5

  binseg_attributes:
    channels_list: [48, 96, 192, 256]
    depths: [3, 9, 3]
    heads_list: [6, 12, 24]
    window_sizes: [0.1, 0.2, 0.4]
    point_embedding:
      use: True
      kpconv_point_influence: 0.015
      kpconv_ball_radius: 0.03
    stratified_union: False
    downsample_ratio: 0.26
    max_drop_path_prob: 0.3
    stratified_downsample_ratio: 0.26
    crpe_bins: 80
    transformer_lr_ratio: 0.1
    checkpoint_path: '/home/mkaailab/Documents/logs/tooth_segmentation/katja_fold3/checkpoints/weights-epoch=199.ckpt'
    lr: 0.0006
    weight_decay: 0.0001
    epochs: 200
    warmup_epochs: 5

  proposal_points: 15000
  dbscan_cfg:
    max_neighbor_dist: 0.03
    min_points: 40  # [30, 50]
    weighted_cluster: True
    weighted_average: True
  do_align: default
  tta: True
```

**Comment by sandroV1972** [comment]
> This appears to be an issue with the segmentation module. You may want to try cleaning the scan or rotating it so that it more closely matches the training data distribution.
> 
> I’ve used his model for inference on many scans, and it generally performs very well.
> 
> If you’re interested, I also have a production-ready intraoral scan segmenter that I trained using Niel’s segmentations as soft labels. It’s built on the Pointcept framework and is extremely fast and reliable for inference. Let me know if that could be helpful.
> 
> Have a nice day 🙂

Hi Matteo @theElandor I have looked at your guthub to see if you already published the repo and could not find it. I tried also to write to you (I could find a Unimore email address) because I am working on something like you are. And I find @nnistelrooij work really helpful. My goal is to retrieve some points and measures from IOS to feed a risk assessment that uses these and some xrays measurements.
Hope to get in touch with you to see if you can help me (both you and Niels.

Cheers
A V

**Comment by theElandor** [comment]
> > This appears to be an issue with the segmentation module. You may want to try cleaning the scan or rotating it so that it more closely matches the training data distribution.
> > I’ve used his model for inference on many scans, and it generally performs very well.
> > If you’re interested, I also have a production-ready intraoral scan segmenter that I trained using Niel’s segmentations as soft labels. It’s built on the Pointcept framework and is extremely fast and reliable for inference. Let me know if that could be helpful.
> > Have a nice day 🙂
> 
> Hi Matteo [@theElandor](https://github.com/theElandor) I have looked at your guthub to see if you already published the repo and could not find it. I tried also to write to you (I could find a Unimore email address) because I am working on something like you are. And I find [@nnistelrooij](https://github.com/nnistelrooij) work really helpful. My goal is to retrieve some points and measures from IOS to feed a risk assessment that uses these and some xrays measurements. Hope to get in touch with you to see if you can help me (both you and Niels.
> 
> Cheers A V

Hi @sandroV1972, I am still working on the model. As soon as it will be published, I will put the repo public :). Reach me via email matteo.lugli@unimore.it for more info if you need.
